test/integration/infrastructure/infrastructure_test.go

```diff
@@ -0,0 +1,561 @@
+}
+
+func prepareNewRouter(ctx context.Context, logger *logrus.Entry, routerName string, openstackClient *OpenstackClient) (*string, error) {
+	logger.Infof("Waiting until router '%s' is created...", routerName)
+
+	createOpts := routers.CreateOpts{
+		Name: routerName,
+	}
+	router, err := routers.Create(openstackClient.NetworkingClient, createOpts).Extract()
+	Expect(err).NotTo(HaveOccurred())
+
+	logger.Infof("Router '%s' is created...", routerName)
+	return &router.ID, nil
+}
+
+func teardownRouter(ctx context.Context, logger *logrus.Entry, routerID string, openstackClient *OpenstackClient) error {
+	logger.Infof("Waiting until router '%s' is deleted...", routerID)
+
+	err := routers.Delete(openstackClient.NetworkingClient, routerID).ExtractErr()
+	Expect(err).NotTo(HaveOccurred())
+
+	logger.Infof("Router '%s' is deleted...", routerID)
+	return nil
+}
+
+type infrastructureIdentifiers struct {
+	networkID  *string
+	keyPair    *string
+	subnetID   *string
+	secGroupID *string
+	routerID   *string
+}
+
+func verifyCreation(
+	ctx context.Context,
+	openstackClient *OpenstackClient,
+	infra *extensionsv1alpha1.Infrastructure,
+	infraStatus *openstackv1alpha1.InfrastructureStatus,
+	providerConfig *openstackv1alpha1.InfrastructureConfig,
+	cidr *string,
+) (infrastructureIdentifier infrastructureIdentifiers) {
+	// router exists
+	router, err := routers.Get(openstackClient.NetworkingClient, infraStatus.Networks.Router.ID).Extract()
+	Expect(err).NotTo(HaveOccurred())
+	Expect(router.Status).To(Equal("ACTIVE"))
+	infrastructureIdentifier.routerID = &router.ID
+
+	// network is created
+	net, err := networks.Get(openstackClient.NetworkingClient, infraStatus.Networks.ID).Extract()
+	Expect(err).NotTo(HaveOccurred())
+	infrastructureIdentifier.networkID = &net.ID
+
+	// subnet is created
+	subnet, err := subnets.Get(openstackClient.NetworkingClient, infraStatus.Networks.Subnets[0].ID).Extract()
+	Expect(err).NotTo(HaveOccurred())
+	Expect(subnet.CIDR).To(Equal(providerConfig.Networks.Workers))
+	infrastructureIdentifier.subnetID = &subnet.ID
+
+	// security group is created
+	secGroup, err := groups.Get(openstackClient.NetworkingClient, infraStatus.SecurityGroups[0].ID).Extract()
+	Expect(err).NotTo(HaveOccurred())
+	Expect(secGroup.Name).To(Equal(infraStatus.SecurityGroups[0].Name))
+	infrastructureIdentifier.secGroupID = &secGroup.ID
+
+	// keypair is created
+	keyPair, err := keypairs.Get(openstackClient.ComputeClient, infraStatus.Node.KeyName).Extract()
+	Expect(err).NotTo(HaveOccurred())
+	infrastructureIdentifier.keyPair = &keyPair.Name
+
+	return infrastructureIdentifier
+}
+
+func verifyDeletion(
+	ctx context.Context,
+	openstackClient *OpenstackClient,
+	infrastructureIdentifier infrastructureIdentifiers,
+	providerConfig *openstackv1alpha1.InfrastructureConfig,
+) {
+	// keypair doesn't exist
+	_, err := keypairs.Get(openstackClient.ComputeClient, *infrastructureIdentifier.keyPair).Extract()
+	Expect(err).To(HaveOccurred())
```

Comment:
Optional suggestion: you could add a check to make sure the error is a 404 error instead of simply an error